### ISSUE TITLE
How to infer prediction type via C API

### ISSUE THREAD

**Issue by travisbrady** [VowpalWabbit/vowpal_wabbit#2416]
## Short description
Looking at the vw Python bindings https://github.com/VowpalWabbit/vowpal_wabbit/blob/master/python/pylibvw.cc#L145 I can see we need to examine the prediction type to know which get prediction method to call.
Currently the C API doesn't seem to expose this (though I'd be happy to be wrong).
So, for instance, if you want a prediction from a CB you need to call `VW_GetCostSensitivePrediction` but you need to know upfront that that is the appropriate method to call as it doesn't seem inferable automatically.

Is there some better way to do this or would it require new code in the C API?

## **How this suggestion will help you/others**
Would make it possible for authors of bindings to the C API to support all of vw's functionality.

## **Possible solution/implementation details**
Add a new function to `vwdll.h` and `vwdll.cpp` (accepting a `VW_EXAMPLE`) to return `all->l->pred_type` and then add constants to `vwdll.h` so binding authors can adjust as appropriate.

**Comment by jackgerrits** [comment]
Yep, I agree. This is very necessary for the C api. This will definitely be incorporated into the API work we are doing.

**Comment by travisbrady** [comment]
Is there any way to do this today though? Or is this blocked on upcoming changes to the C API?

**Comment by jackgerrits** [comment]
There isn't as far as I can tell. It seems to me that there aren't entry points to retrieve all of the different prediction types either. We could patch the existing API, even thought it will be deprecated soon. @peterychang thoughts?

**Comment by peterychang** [comment]
I think patching the existing API is fine as a quick fix. If possible, we should make the signature as close to the proposed APIs as we can though.

@travisbrady Thanks for the feedback so far. I've migrated the proposed API surface to the wiki https://github.com/VowpalWabbit/vowpal_wabbit/wiki/%5BWIP%5D-C-API-Proposal
Its still a work in progress, but if you want to take an look and have any comments or concerns we'll be happy to address them.

**Comment by JohnLangford** [comment]
There's a good idea here, and a patch is welcome.
